src/core/cluster.py

```diff
@@ -4,30 +4,53 @@
 
 """Collection of global cluster state for the ZooKeeper quorum."""
 import logging
-from typing import Set
-
+from typing import Dict, Set
+
+from charms.data_platform_libs.v0.data_interfaces import (
+    DatabaseProviderData,
+    DataPeerData,
+    DataPeerOtherUnitData,
+    DataPeerUnitData,
+)
 from ops.framework import Framework, Object
-from ops.model import Relation
+from ops.model import Relation, Unit
 
 from core.models import SUBSTRATES, ZKClient, ZKCluster, ZKServer
-from literals import CLIENT_PORT, PEER, REL_NAME, SECURE_CLIENT_PORT, Status
+from literals import (
+    CLIENT_PORT,
+    PEER,
+    REL_NAME,
+    SECRETS_UNIT,
+    SECURE_CLIENT_PORT,
+    Status,
+)
 
 logger = logging.getLogger(__name__)
 
 
 class ClusterState(Object):
-    """Collection of global cluster state for the ZooKeeper quorum."""
+    """Collection of global cluster state for Framework/Object."""
 
     def __init__(self, charm: Framework | Object, substrate: SUBSTRATES):
         super().__init__(parent=charm, key="charm_state")
         self.substrate: SUBSTRATES = substrate
 
+        self.peer_app_interface = DataPeerData(self.model, relation_name=PEER)
+        self.peer_unit_interface = DataPeerUnitData(
+            self.model, relation_name=PEER, additional_secret_fields=SECRETS_UNIT
+        )
+        self.client_provider_interface = DatabaseProviderData(self.model, relation_name=REL_NAME)
+
+        self._servers_data = {}
+
     # --- RAW RELATION ---
 
     @property
-    def peer_relation(self) -> Relation | None:
+    def peer_relation(self) -> Relation:
         """The cluster peer relation."""
-        return self.model.get_relation(PEER)
+        if not (peer_relation := self.model.get_relation(PEER)):
+            raise AttributeError(f"No peer relation {PEER} found.")
+        return peer_relation
```

Comment:
I'm not sure raising here is right. There are some checks around that directly look for peer relation existing i.e.:  `if not self.peer_relation: defer()`. Raising renders those checks unusable, because this property now only has two exit cases:
- Raise AttributeError if the relation doesn't exist.
- Return the relation.

Reply:
I don't insist on raising -- but I'm afraint `None` won't work here either.

Let me get back to the precise issue why I went with this solution.

The list of issues this is solving was longer than the issues it's introducing. More news later.

Reply:
If `cached_property` is not 100% guarantee, I woudln't risk. I'll check (never used it before).

Reply:
OK, I remembered the reason for this decision. And I still stand by it.
I believe we need to further extend the solution to situations such as external relations that may or may not be there (untlie `peer`) 

This is NOT about linting.

Without ensuring the existence of the relation, we impose severe risk of VERY fragile code.

Detailed explanation [here](https://github.com/canonical/kafka-operator/pull/189/files#r1568965741)

Reply:
@zmraul  The case you describe has been investigated carefully. There is a single case in ZK where peer relation may be `None` (and we `defer()`). This is the reason why the `has_peer_relation()` query was introduced and is being used.

Similar security measure can be simply applied on similar cases. Instead of

```
if not self.state.cluster.peer_relation:
    event.defer()
    return
```
we should do
```
if not self.state.cluster.has_peer_relation():
    event.defer()
    return
```

It's not a major change :-)

Reply:
Implemented as required.